enhancements/network/routed-ingress-support-for-primary-udn-attached-vms-with-static-ips.md

```diff
@@ -0,0 +1,737 @@
+attribute to request specific IPs for the workload which will consume the
+claim.
+
+Below you'll see the changes requested:
+```go
+type IPAMClaimSpec struct {
+	// The network name for which this persistent allocation was created
+	Network string `json:"network"`
+	// The pod interface name for which this allocation was created
+	Interface string `json:"interface"`
++	// The IPs requested by the user
++	// +optional
++	IPRequests []CIDR           `json:"ipRequests,omitempty"`
++	MACRequest net.HardwareAddr `json:"macRequest,omitempty"`
+}
+
+// IPAMClaimStatus contains the observed status of the IPAMClaim.
+type IPAMClaimStatus struct {
+    // The list of IP addresses (v4, v6) that were allocated for the pod interface
+-   IPs []string `json:"ips"`
++   IPs []CIDR `json:"ips"`
++   // The name of the pod holding the IPAMClaim
++   OwnerPod string `json:"ownerPod"`
++   // Conditions contains details for one aspect of the current state of this API Resource
++   Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+```
+
+The `IPAMClaim` status will have (at least) the following conditions:
+- SuccessfulAllocation: reports the IP address was successfully allocated for
+  the workload
+- AllocationConflict: reports the requested allocation was not successful - i.e.
+  the requested IP address is already present in the network
+
+These conditions will look like - **successful** allocation:
+```
+type: SuccessfulAllocation
+status: "True"
+reason: IPAllocated
+message: "IP 192.168.1.5 allocated successfully"
+lastTransitionTime: "2025-05-13T11:56:00Z"
+```
+
+These conditions will look like - **failed** allocation:
+```
+type: AllocationConflict
+status: "True"
+reason: IPAlreadyExists
+message: "Requested IP 192.168.1.5 is already assigned in the network"
+lastTransitionTime: "2025-05-13T11:56:00Z"
+```
+
+#### New IPPool CRD
+
+**NOTE:** this CRD is only required if we go with a
+[centralized IP allocation](#centralized-ip-management) design.
+
+The IPPool CRD will operate as a place to store the MAC to IP addresses
+association for a logical network.
+
+```go
+type IPPool struct {
+	metav1.TypeMeta     `json:",inline"`
+	metav1.ObjectMeta   `json:"metadata,omitempty"`
+
+	Spec   IPPoolSpec   `json:"spec,omitempty"`
+	Status IPPoolStatus `json:"status,omitempty"`
+}
+
+type IPPoolSpec struct {
+	NetworkName string                         `json:"network-name"`
+	Entries map[net.HardwareAddr][]net.IPNet   `json:"entries"`
+}
+
+type IPPoolStatus struct {
+	Conditions []Condition      `json:"conditions"`
+	// this is an optional improvement for improving the UX
+	AssociatedNADs []NADInfo    `json:"associated-nads,omitempty"`
+}
+
+type NADInfo struct {
+	Name string `json:"name"`
+}
+```
+
+The `IPPool` CRD will have at least the following conditions:
+- DuplicateMACAddresses: will indicate to the admin that a MAC address appears
+  multiple times in the `Entries` list
```

Comment:
Will the DuplicateMACAddresses block VM creation?
Bit worried about cross namespace migrations.

Reply:
Yes, but unless said VMs already had duplicate MACs on the source cluster, they won't have duplicate MACs in the dst cluster, right ? 

Keep in mind we're (kind of ...) attempting to **re-create** the network that existed in the source cluster - we don't have L2 connectivity to it.